rfcs/20221006-frontend_event_tracking.md

```diff
@@ -0,0 +1,294 @@
+  };
+  /** The user uses the 'back to search' link on a single result **/
+  BACK_TO_SEARCH: {
+    /** The unique ID of the media */
+    id: string;
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+  /** The visibility of the filter sidebar on desktop is toggled **/
+  TOGGLE_FILTERS: {
+    /** The media type being searched */
+    mediaType: MediaType;
+  };
+};
+type AnalyticsEventTypes = keyof AnalyticsEvents;
+
+/**
+ * Send an event to our analytics endpoint. Should be a no-op if the
+ * user has opted-out of analytics.
+ */
+async function sendEvent<T extends AnalyticsEventTypes>(
+  key: T,
+  payload?: AnalyticsEvents[T]
+): Promise<any> {
+  /**
+   * The default values to send with every analytics event.
+   */
+  const defaults = {
+    timestamp: Date.now(),
+    path: window.location.pathname,
+    // The dimensions of the user's device
+    width: window.innerWidth,
+    height: window.innerHeight,
+    referrer: document.referrer || null,
+  };
+  /**
+   * Actually send the event here, with some additional values.
+   * I would have prefered to use `navigator.sendBeacon` here
+   * but it is blocked by most ad blockers.
+   *
+   * We might not want to call our api route "analytics" to
+   * further avoid blocking by adblockers, but I'll use that
+   * name for clarity here.
+   */
+  const response = await fetch(`/analytics`, {
+    method: "POST",
+    headers: { "Content-Type": "application/json" },
+    body: JSON.stringify({ ...defaults, ...payload }),
+  });
+}
+```
+
+### Implementation Plan
+
+This needs more elaboration, as it's a very rough outline. For example, specific events will require more detailed planning.
+
+- [ ] Create "Event Tracking" GitHub milestone in the frontend repository
+- [ ] Design: design a privacy and analytics opt-out checkbox on the Privacy page
+- [ ] Create an issue for the `ANALYTICS_ENABLED` feature flag
+- [ ] Create an issue for the creation of the 'sendEvent' function and for the api route to post analytics events to
+- [ ] Create a single issue for each event in the sample implementation. Each issue will need its specific implementation (when the event is called, what user DOM interaction triggers it, etc.) approved by frontend developers before it is implemented.
+
+## Concerns / Pitfalls
+
+- Related to content safety, this is an area where we are surfacing user information in the form of text, or even malicious URLs, to Openverse developers. These searches could include explicit or sensitive material. We're also potentially surfacing explicit image results to anyone working with this data.
+- I think it's great to build this decoupled from a specific back-end, but if we never actually implement a backend this would largely be a wasted effort.
+- We're building something here that is commonly drop-in when 3rd party tools are used. At least for page views; custom events still require custom work. Is this a poor use of time?
+
+### Prior Art
```

Comment:
Did you take a look at GoatCounter, by any chance? It has a Django middleware plugin that makes it pretty easy to get analytics from our API usage as well, not just from the frontend: https://www.goatcounter.com/

Implementing it would be pretty trivial as well (in my experience) and it supports data attribute based event tracking. It allows you to easily configure click events by assigning attributes to the elements that might be clicked on. It also supports an event function API similar to the one you described that we could use and you can send events directly to the API endpoint too so we could proxy if it we do indeed want to do that.

I've used it for a personal project and the dashboard is also quite nice to use.

One question I have for any of these providers though is whether we're able to easily get the data out so that we can do analysis on it ourselves in the future if we outgrow the capabilities of the tool itself. In particular if we started deriving scores for results based on how likely they are to be clicked on when they appear in a search, we would need to be able to ETL the data out of the analytics provider and into our own data warehouse. If the tool does _not_ support that then IMO it is a non-starter.

Reply:
I may be misunderstanding what you mean by custom events, but I believe it does support them, at least in my understanding of them, I use them in my personal project: https://www.goatcounter.com/help/events#sending-events-from-javascript-44

It handles page views automatically but click events can be done either via JS (linked directly above) or via the data attributes I mentioned.

Reply:
Self hosting GoatCounter is pretty easy too, if we want to go that route. If it's something we'd want to do, it'd be nice to evaluate the different providers to see what the effort should be. I'd be surprised if any are much more than a single docker container rigged up to a database though.

Reply:
I can clarify what I mean about the events. My language was definitely imprecise so apologies.

As fas as I could tell GoatCounter's "javascript events" are still _meant to_ be page view events. Based on the docs they have a strict payload signature that only allows for an ID, title, and referrer for each event. I think we would benefit from arbitrary payloads (enforced via types for each kind of event) to allow us to ask some pretty specific questions of our data. GoatCounter's events can also only be triggered by click, which while is a good way of detecting user interactions is also a bit too limiting, IMO. For example, say we wanted to keep track of how long users spend on the page, we'd want our event to be triggered by the `beforeunload` event, or if we wanted to trigger an event based on right clicks (as one event in my proposal does), we wouldn't be able to do it with GoatCounter.

Reply:
> GoatCounter's events can also only be triggered by click

This isn't true, you can send events from JavaScript on arbitrary actions (re-linking the doc I shared above): https://www.goatcounter.com/help/events#sending-events-from-javascript-44

> Based on the docs they have a strict payload signature that only allows for an ID, title, and referrer for each event

Gotcha! That does seem to be a disqualifying limitation, good call!